### PR TITLE
Trusted badges, out-of-band avatars, replies and forwards UI updated, new websocket.

### PR DESCRIPTION
**Pull request by or-else**
Everything should work more or less although some bugs likely exist.

### CODE REVIEW

File: TinodeSDK/Connection.swift [hunk 1]
```diff
@@ -2,13 +2,12 @@
 //  Connection.swift
 //  ios
 //
-//  Copyright © 2019 Tinode. All rights reserved.
+//  Copyright © 2019-2022 Tinode LLC. All rights reserved.
 //
 
 import Foundation
-import SwiftWebSocket
```
Comment (aforge) [inline TinodeSDK/Connection.swift]: Nice!

File: TinodeSDK/Tinode.swift [hunk 2]
```diff
@@ -1201,20 +1194,10 @@ public class Tinode {
...
-        return head
-    }
-
-    public func publish(topic: String, head: [String: JSONValue]?, content: Drafty) -> PromisedReply<ServerMessage> {
+    public func publish(topic: String, head: [String: JSONValue]?, content: Drafty, attachments: [String]?) -> PromisedReply<ServerMessage> {
         let msgId = getNextMsgId()
-        let msg = ClientMessage<Int, Int>(
-            pub: MsgClientPub(id: msgId, topic: topic, noecho: true, head: head, content: content))
+        let msg = ClientMessage<Int, Int>(pub: MsgClientPub(id: msgId, topic: topic, noecho: true, head: head, content: content))
+        msg.extra = MsgClientExtra(attachments: attachments)
```
Comment (aforge) [inline TinodeSDK/Tinode.swift]: Consider checking if `attachments` is not empty:
```
if let attachments = attachments, !attachments.isEmpty {
  ...
}
```

File: TinodeSDK/Topic.swift [hunk 3]
```diff
@@ -1188,13 +1227,18 @@ open class Topic<DP: Codable & Mergeable, DR: Codable & Mergeable, SP: Codable,
...
-        if let extra = extraHeaders {
-            head = head ?? [:]
-            // Break conflicts by taking values from extras.
-            head!.merge(extra) { (_, new) in new }
+        var head = extraHeaders
+        if !content.isPlain {
+            if head == nil {
+                head = [:]
+            }
+            head!["mime"] = JSONValue.string(Drafty.kMimeType)
```
Comment (aforge) [inline TinodeSDK/Topic.swift]: Maybe add a comment stating that the Tinode SDK only sets the `mime` header, the other headers are set by the caller.

File: TinodeSDK/WebSocket.swift [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+    private func listen()  {
+        guard socket.state == .running else { return }
+
+        socket.receive { [weak self] result in
+            guard let self = self else { return }
+
+            switch result {
+            case .failure(let error):
+                self.delegate?.onError(connection: self, error: error)
```
Comment (aforge) [inline TinodeSDK/WebSocket.swift]: This won't call `self.listen()`. That is, if we get a socket error (e.g. network failure), the Websocket object will stop operating. Is it intentional?

File: TinodeSDK/model/Subscription.swift [hunk 5]
```diff
@@ -141,27 +153,27 @@ public class Subscription<SP: Codable, SR: Codable>: SubscriptionProto {
         if sub.acs != nil {
             if acs == nil {
                 self.acs = Acs(from: sub.acs!)
-                changed += 1
+                changed = true
             } else {
-                changed += (acs!.merge(from: sub.acs) ? 1 : 0)
+                changed = acs!.merge(from: sub.acs)
```
Comment (aforge) [inline TinodeSDK/model/Subscription.swift]: Should be
```
changed = acs!.merge(from: sub.acs) || changed
```

File: TinodeSDK/model/Subscription.swift [hunk 6]
```diff
@@ -171,18 +183,18 @@ public class Subscription<SP: Codable, SR: Codable>: SubscriptionProto {
...
+            changed = true
         }
         if sub.seen != nil {
             if seen == nil {
                 seen = sub.seen
-                changed += 1
+                changed = true
             } else {
-                changed += (seen!.merge(seen: sub.seen) ? 1 : 0)
+                changed = seen!.merge(seen: sub.seen)
```
Comment (aforge) [inline TinodeSDK/model/Subscription.swift]: Same here. `changed = seen.merge() || changed`

File: TinodeSDK/model/Types.swift [hunk 7]
```diff
@@ -2,61 +2,86 @@
...
+        return nil
+    }
+
     public var comment: String? {
         get {
-            if case let .string(v)? = self["comment"] {
-                return v
-            }
-            return nil
+            return getStringValue(name: "comment")
```
Comment (aforge) [inline TinodeSDK/model/Types.swift]: This can be done even easier:
```
return self[name]?.asString()
```

**Review by aforge — APPROVED** [review]

File: TinodeSDK/Tinode.swift (hunk 2 above)
Comment (or-else) [reply TinodeSDK/Tinode.swift]: done

File: TinodeSDK/model/Subscription.swift (hunk 5 above)
Comment (or-else) [reply TinodeSDK/model/Subscription.swift]: done

File: TinodeSDK/model/Subscription.swift (hunk 6 above)
Comment (or-else) [reply TinodeSDK/model/Subscription.swift]: done

File: TinodeSDK/model/Types.swift (hunk 7 above)
Comment (or-else) [reply TinodeSDK/model/Types.swift]: done

File: TinodeSDK/Topic.swift (hunk 3 above)
Comment (or-else) [reply TinodeSDK/Topic.swift]: done

File: TinodeSDK/WebSocket.swift (hunk 4 above)
Comment (or-else) [reply TinodeSDK/WebSocket.swift]: It was a bug. Fixed now.

**Pull request merged**